estimagic/differentiation/generate_steps.py

```diff
@@ -0,0 +1,255 @@
+    to the side that has more space to the bound if there is a bound violation.
+
+    Args:
+        x (np.ndarray): 1d array with parameters.
+        pos (np.ndarray): Array with positive steps of shape (n_steps, len(x))
+        neg (np.ndarray): Array with negative steps of shape (n_steps, len(x))
+        method (str): One of ["forward", "backward"]
+        lower_step_bounds (np.ndarray): Lower bounds for steps.
+        upper_step_bounds (np.ndarray): Upper bounds for steps.
+
+    Returns:
+        pos (np.ndarray): Copy of pos with additional NaNs
+        neg (np.ndarray): Copy of neg with additional NaNs
+
+    """
+    pos = pos.copy()
+    neg = neg.copy()
+    better_side = np.where(upper_step_bounds >= -lower_step_bounds, np.ones_like(x), -1)
+    max_abs_step = pos[:, -1]
+    if method == "forward":
+        used_side = np.where(
+            upper_step_bounds >= max_abs_step, np.ones_like(x), better_side
+        )
+    elif method == "backward":
+        used_side = np.where(
+            -lower_step_bounds >= max_abs_step, -np.ones_like(x), better_side
+        )
+    else:
+        raise ValueError("This function only works for forward or backward method.")
+
+    pos[used_side == -1] = np.nan
+    neg[used_side == 1] = np.nan
+    return pos, neg
+
+
+def _rescale_to_accomodate_bounds(
+    base_steps, pos, neg, lower_step_bounds, upper_step_bounds, min_steps
+):
+    """Rescale steps to make them compatible with bounds unless this violates min_steps.
+
+    Args:
+        base_steps (np.ndarray, optional): 1d array of the same length as x.
+            base_steps * scaling_factor is the absolute value of the first (and possibly
+            only) step used in the finite differences approximation of the derivative.
+        pos (np.ndarray): Array with positive steps of shape (n_steps, len(x))
+        neg (np.ndarray): Array with negative steps of shape (n_steps, len(x))
+        lower_step_bounds (np.ndarray): Lower bounds for steps.
+        upper_step_bounds (np.ndarray): Upper bounds for steps.
+        min_steps (np.ndarray): Minimal possible step sizes that can be chosen
+            to accomodate bounds. Needs to have same length as x.
+
+    Returns:
+        pos (np.ndarray): Copy of pos with rescaled steps.
+        neg (np.ndarray): Copy of neg with rescaled steps.
+
+    """
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=RuntimeWarning)
+        pos_needed_scaling = _fillna(
+            upper_step_bounds / np.nanmax(pos, axis=1), 1
+        ).clip(0, 1)
+        neg_needed_scaling = _fillna(
+            lower_step_bounds / np.nanmin(neg, axis=1), 1
+        ).clip(0, 1)
+    needed_scaling = np.minimum(pos_needed_scaling, neg_needed_scaling)
+
+    min_possible_scaling = min_steps / base_steps
+
+    scaling = np.maximum(needed_scaling, min_possible_scaling).reshape(-1, 1)
+
+    pos = pos * scaling
+    neg = neg * scaling
+    return pos, neg
```

Comment:
In the docstring you say "[2. decrease the steps if possible]. 3. Set the conflicting steps to NaN ...". 
I don't think you actually do step 3 here.

Reply:
I think I am doing it [here](https://github.com/OpenSourceEconomics/estimagic/blob/33b20dfe69a1fa950162a06f8ccebb7494eac88f/estimagic/differentiation/generate_steps.py#L117)

Reply:
yes, that's right!